### PR TITLE
[feat:#17][tsdb:index]index write and read base on B+Tree

### PR DESCRIPTION
**Pull request by lizunmvn**

### CODE REVIEW

**Comment by CodingCrush** [comment]
rename it to `[tsdb:index]`

File: pkg/buf/buf.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import (
+	"bytes"
+	"encoding/binary"
+)
+
+var scratch [binary.MaxVarintLen64]byte
+
+//=================================buf writer
+//Create a ByteBuff
+func NewByteBuff() *bytes.Buffer {
```
Comment (stone1100) [inline pkg/buf/buf.go]: maybe need refactor, can use pkg/stream/binary

File: pkg/constants/index_constants.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package constants
+
+import (
+	"encoding/json"
+	"math"
+)
+
+const (
```
Comment (stone1100) [inline pkg/constants/index_constants.go]: put it under the package "tsdb/index"

File: pkg/util/sort.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package util
```
Comment (stone1100) [inline pkg/util/sort.go]: using "sort" package?

File: tsdb/index/field_uid.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	return nil
+}
+
+//Forces a flush of in-memory data, and clear it
+func (f *FieldUID) Flush() error {
+	//sort metricIds
+	metricIds := f.getSortMetricIds()
+	//flush fieldId to kv-store
+	flusher := f.family.NewFlusher()
+	f.rwLock.Lock()
```
Comment (stone1100) [inline tsdb/index/field_uid.go]: refactor using "COW"

Comment (stone1100) [inline tsdb/index/field_uid.go]: Lock granularity is too large

File: pkg/constants/index_constants.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+func StringToMap(tags string) map[string]string {
+	var tagsMap map[string]string
+	_ = json.Unmarshal([]byte(tags), &tagsMap)
+	return tagsMap
+}
+
+func MapToString(tagsMap map[string]string) string {
+	b, _ := json.Marshal(tagsMap)
```
Comment (stone1100) [inline pkg/constants/index_constants.go]: need handle error

File: pkg/util/bytes.go [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+func ShortToInt(high uint16, low uint16) uint32 {
+	return uint32(high)<<16 + uint32(low)
+}
+
+func IntToShort(value uint32) (high, low uint16) {
+	high = uint16(value >> 16 & math.MaxUint16)
+	low = uint16(value & math.MaxUint16)
+	return
+}
```
Comment (CodingCrush) [inline pkg/util/bytes.go]: these functions belongs to  `index`

File: pkg/util/sort.go (hunk 3 above)
Comment (CodingCrush) [reply pkg/util/sort.go]: duplicate to `sort.Slice`

File: tsdb/index/field_uid.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	return &FieldUID{
+		metricMap: make(map[uint32]*FieldIndex),
+		family:    f,
+	}
+}
+
+//Finds the ID associated with a given field name or creates it.
+func (f *FieldUID) GetOrCreateFieldID(metricID uint32, fieldName string, fieldType field.Type,
+	create bool) (uint32, error) {
+	fieldIndex := f.metricMap[metricID]
```
Comment (CodingCrush) [inline tsdb/index/field_uid.go]: not thread-safe

`fieldIndex, ok := f.metricMap[metricID]` is more Intuitive

File: tsdb/index/metric_uid.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	}
+	return constants.NotFoundMetricID, false
+}
+
+func (m *MetricUID) SuggestMetrics(prefix string, limit int16) map[string]struct{} {
+	if len(prefix) > 0 {
+		nameBytes := []byte(prefix)
+		partition := getPartition(nameBytes)
+		snapshot, err := m.family.GetSnapshot(partition)
+		if nil == err {
```
Comment (CodingCrush) [inline tsdb/index/metric_uid.go]: Nesting level too deep!
how about use this
`
if err!= nil {
return nil
}
if readers == nil{
return nil
}
`

File: tsdb/uid.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	GetOrCreateMetricID(metricName string, create bool) (uint32, bool)
+	//Attempts to find suggestions of metric names given a search prefix.
+	SuggestMetrics(prefix string, limit int16) map[string]struct{}
+	//Forces a flush of in-memory data, and clear it
+	Flush() error
+}
+
+//TagsUID is unique under the metric name.
+//Shard level sharing.
+type TagsUID interface {
```
Comment (CodingCrush) [inline tsdb/uid.go]: the definition is TSID?

File: tsdb/index/field_uid.go (hunk 4 above)
Comment (lizunmvn) [reply tsdb/index/field_uid.go]: Remove lock dependencies

Comment (lizunmvn) [reply tsdb/index/field_uid.go]: Remove lock dependencies, multithreading is not safe

File: pkg/util/sort.go (hunk 3 above)
Comment (lizunmvn) [reply pkg/util/sort.go]: Sort.Slice does not support uint32

**Pull request merged**